thrust/examples/run_length_decoding_single_pass.cu

```diff
@@ -0,0 +1,169 @@
+      : value(value)
+      , count(count)
+      , offset(offset)
+      , run_id(run_id)
+  {}
+  run(run const& other)            = default;
+  run& operator=(run const& other) = default;
+
+  __host__ __device__ friend run operator+(run l, run r)
+  {
+    return run{r.value, r.count, l.offset + l.count + r.offset, l.run_id + r.run_id};
+  }
+};
+
+template <typename OutputIterator, typename CountType, typename ExpandedSizeIterator>
+struct expand
+{
+  OutputIterator out;
+  CountType out_size;
+  CountType runs_size;
+  ExpandedSizeIterator expanded_size;
+
+  __host__ __device__
+  expand(OutputIterator out, CountType out_size, CountType runs_size, ExpandedSizeIterator expanded_size)
+      : out(out)
+      , out_size(out_size)
+      , runs_size(runs_size)
+      , expanded_size(expanded_size)
+  {}
+
+  template <typename ValueType>
+  __host__ __device__ CountType operator()(run<ValueType, CountType> r) const
+  {
+    cuda::std::size_t end = cuda::minimum()(r.offset + r.count, out_size);
+    thrust::fill(thrust::seq, out + r.offset, out + end, r.value);
+    if (r.run_id == runs_size) // If we're the last run, write the expanded size.
+    {
+      *expanded_size = end;
+    }
+    return end;
+  }
+};
+
+template <typename ValueIterator, typename CountIterator, typename OutputIterator, typename CountType>
+OutputIterator run_length_decode(
+  ValueIterator values, CountIterator counts, CountType runs_size, OutputIterator out, CountType out_size)
+{
+  using ValueType = typename ValueIterator::value_type;
+
+  // Zip the values and counts together and convert the resulting tuples to a named struct.
+  auto runs = thrust::make_transform_iterator(
+    thrust::make_zip_iterator(values, counts), [] __host__ __device__(thrust::tuple<ValueType, CountType> tup) {
+      return run{thrust::get<0>(tup), thrust::get<1>(tup)};
+    });
+
+  // Allocate storage for the expanded size. If we were using CUB directly, we could read this
+  // from the final ScanTileState.
+  thrust::device_vector<CountType> expanded_size(1);
+
+  // Iterator that writes the expanded sequence to the output and discards the actual scan results.
+  auto expand_out = thrust::make_transform_output_iterator(
+    thrust::make_discard_iterator(), expand(out, out_size, runs_size, expanded_size.begin()));
+
+  // Scan to compute output offsets and then write the expanded sequence.
+  thrust::inclusive_scan(thrust::device, runs, runs + runs_size, expand_out);
+
+  // Calculate the end of the output sequence.
+  return out + expanded_size[0];
+}
+
+int main()
+{
+  using ValueType = cuda::std::int32_t;
+  using CountType = cuda::std::size_t;
+
+  CountType size   = 8192;
+  CountType repeat = 4;
+
+  thrust::device_vector<ValueType> values(size);
+  thrust::sequence(thrust::device, values.begin(), values.end(), 0);
+
+  thrust::device_vector<CountType> counts(size);
+  thrust::fill(thrust::device, counts.begin(), counts.end(), repeat);
```

Comment:
This should rather be 
```suggestion
  thrust::device_vector<CountType> counts(size, repeat);
```

Reply:
> I prefer to be explicit. The device_vector constructor doesn't make it as clear what's going on.

While this may be a somewhat valid argument, the code now writes to memory twice. The vector constructor zeros it, then the fill overwrites all the zeros. I think this is bad practice as well.